[CASSANDRA-16378] Expose all client options via system_views.clients and nodetool clientstats

rtib opened the pull request.

blerer commented on src/java/org/apache/cassandra/transport/ConnectedClient.java:
```diff
@@ -128,10 +137,20 @@ private SslHandler sslHandler()
+        } catch(JsonProcessingException e)
+        {
+            clientOptionsJSON = UNDEFINED;
+        }
```
We do not normally use json libraries for that type of things. It is also more readable to us Guava Joiner:
` Joiner.on(',').withKeyValueSeparator("=").join(clientOptions().get())`

blerer reviewed (commented):
We do not have the mean to really test the new options but we should be able to partially test the output of the `ClientsTable` queries by verifying only the part that we can predict. Could you create a unit test similar to `SystemPropertiesTableTest`? The call to `executeNet` will open one connection to the server.

rtib commented:
I've added a ClientsTableTest which is querying the virtual table and asserting some values.

First I've added assertion for driver_name and driver_version columns, which values should be contained in the client_options, however, this depends on the behaviour of StartupMessage rather than Clients itself, thus testing this is violating the SUT boundary. I don't know if it is a good idea to keep those ones.

blerer commented on test/unit/org/apache/cassandra/db/virtual/ClientsTableTest.java:
```diff
@@ -0,0 +1,70 @@
+        for (Row r : result)
+        {
+            Assert.assertEquals(r.getInet("address"), InetAddress.getLoopbackAddress());
+            Assert.assertNotNull(r.getInt("port"));
+            Assert.assertTrue(r.getInt("port") > 0);
+            Assert.assertNotNull(r.getMap("client_options", String.class, String.class));
+            // the following are questionable if they belong here
+            Assert.assertEquals(r.getString("hostname"), "localhost");
+            Assert.assertEquals(r.getString("driver_name"), r.getMap("client_options", String.class, String.class).get("DRIVER_NAME"));
+            Assert.assertEquals(r.getString("driver_version"), r.getMap("client_options", String.class, String.class).get("DRIVER_VERSION"));
```
The assertions surrounding the map can be be simplified by using AspectJ assertions (`org.assertj.core.api.Assertions`). Like: `assertThat(map).hasEntrySatisfying(...)`
I think that we can also test `request_count`.

blerer reviewed (commented):
Nice work :-) I just have some minor nits.

rtib commented:
Asserting request_count to some positive, as it has different values in the two rows. Changed map assertion as suggested.

rtib commented:
Squashed and rebased to current trunk.

ekaterinadimitrova2 commented on test/unit/org/apache/cassandra/db/virtual/ClientsTableTest.java:
```diff
@@ -0,0 +1,78 @@
+            Assert.assertTrue(r.getLong("request_count") > 0 );
+            // the following are questionable if they belong here
+            Assert.assertEquals("localhost", r.getString("hostname"));
+            Assertions.assertThat(r.getMap("client_options", String.class, String.class))
+                    .hasEntrySatisfying("DRIVER_VERSION", value -> {
+                        assertThat(value.contains(r.getString("driver_name")));
+                    })
+                    .hasEntrySatisfying("DRIVER_VERSION", value -> {
+                        assertThat(value.contains(r.getString("driver_version")));
+                    });
```
nit: I will probably write those as follows:
```suggestion
            Assertions.assertThat(r.getMap("client_options", String.class, String.class))
                    .hasEntrySatisfying("DRIVER_VERSION", value -> assertThat(value.contains(r.getString("driver_name"))))
                    .hasEntrySatisfying("DRIVER_VERSION", value -> assertThat(value.contains(r.getString("driver_version"))));
```

bereng commented on test/unit/org/apache/cassandra/db/virtual/ClientsTableTest.java:
```diff
@@ -0,0 +1,74 @@
+/*
```
In `ToolRunner` you have methods to invoke `nodetool` so you could add a test for `nodetool` as well. If that doesn't cut it a dtest could imo.

bereng replied:
I wanted to add more info. If you look at `ToolRunner.invokeNodetool()` callers you'll get plenty of examples of such tests. You will see there are methods to check return codes, std output, errors if any are expected, etc. `TableStatsTest` could be used as a guide. But feel free to do your take on it.

The important bit here is that test will check i.e.: The tool doesn't explode when using that option, the new info is indeed returned, return codes are as expected, tests things from the same top-level point of view a user will, etc.

Please let me know if you find any trouble getting it to work and I can happily lend a hand.

bereng commented on src/java/org/apache/cassandra/transport/ConnectedClient.java:
```diff
@@ -132,6 +140,8 @@ private SslHandler sslHandler()
                            .put(ADDRESS, remoteAddress().toString())
                            .put(USER, username().orElse(UNDEFINED))
                            .put(VERSION, String.valueOf(protocolVersion()))
+                           .put(CLIENT_OPTIONS, Joiner.on(',').withKeyValueSeparator('=').join(
```
I would join on `, `. The extra space makes things easier to read on the eyes.

bereng commented on test/unit/org/apache/cassandra/tools/nodetool/ClientStatsTest.java:
```diff
@@ -0,0 +1,130 @@
+                        "        --by-protocol\n" +
+                        "            Lists most recent client connections by protocol version\n" +
+                        "\n" +
+                        "        --clear-history\n" +
```
Nice test!. If you add a test case for `--clear-history`, despite not being directly related to the scope of this ticket, you'd have a complete `clientstats` test class which I think it's a quick win. Wdyt?

bereng commented:
Hi I was wondering about the revert of the test?

rtib commented:
Just moved that into another branch. I'd like to fix it, but currently I didn't found how.

ekaterinadimitrova2 commented on src/java/org/apache/cassandra/transport/ConnectedClient.java:
```diff
@@ -132,6 +140,8 @@ private SslHandler sslHandler()
                            .put(USER, username().orElse(UNDEFINED))
                            .put(VERSION, String.valueOf(protocolVersion()))
+                           .put(CLIENT_OPTIONS, Joiner.on(", ").withKeyValueSeparator('=').join(
+                                   clientOptions().orElse(Collections.<String, String>emptyMap())))
```
```suggestion
                           .put(CLIENT_OPTIONS, Joiner.on(", ")
                                                                             .withKeyValueSeparator("=")
                                                                             .join(clientOptions().orElse(Collections.emptyMap())))
```

ekaterinadimitrova2 commented on test/unit/org/apache/cassandra/db/virtual/ClientsTableTest.java:
```diff
@@ -0,0 +1,74 @@
+        for (Row r : result)
+        {
+            Assert.assertEquals(InetAddress.getLoopbackAddress(), r.getInet("address"));
+            Assert.assertNotNull(r.getInt("port"));
```
Seems like `getInt` already checks for null?

ekaterinadimitrova2 commented on test/unit/org/apache/cassandra/db/virtual/ClientsTableTest.java:
```diff
@@ -67,12 +67,8 @@ public void testSelectAll() throws Throwable
-                        assertThat(value.contains(r.getString("driver_version")));
-                    });
+                    .hasEntrySatisfying("DRIVER_VERSION", value -> assertThat(value.contains(r.getString("driver_name"))))
+                    .hasEntrySatisfying("DRIVER_VERSION", value -> assertThat(value.contains(r.getString("driver_version"))));
```
nit: formatting is a bit off

ekaterinadimitrova2 commented on test/unit/org/apache/cassandra/tools/nodetool/ClientStatsTest.java:
```diff
@@ -0,0 +1,162 @@
+    @Before
+    public void config() throws Throwable
+    {
+        ResultSet result = executeNet("select release_version from system.local");
```
This seems unused?

ekaterinadimitrova2 commented on test/unit/org/apache/cassandra/tools/nodetool/ClientStatsTest.java:
```diff
@@ -0,0 +1,162 @@
+    }
+    
+    @Test
+    public void testClientStatsByProtocol() throws Throwable
```
These `Throwables` are not needed

ekaterinadimitrova2 reviewed (commented):
LGTM, I added a few comments that I also suggest on this branch https://github.com/apache/cassandra/compare/trunk...ekaterinadimitrova2:CASSANDRA-16378

rtib committed:
[CASSANDRA-16378] Expose all client options via system_views.clients and nodetool clientstats


Update test/unit/org/apache/cassandra/db/virtual/ClientsTableTest.java

Co-authored-by: ekaterinadimitrova2 <ebd18@pitt.edu>
add space to join options on


add some basic tests for nodetool clientstats


add clear connection history test


Revert "add clear connection history test"

This reverts commit 4bab747709c8bd355e763b4b7dbae4d776bf6010.

nodetool clientstats --clear-history test


Update src/java/org/apache/cassandra/transport/ConnectedClient.java

Co-authored-by: ekaterinadimitrova2 <ebd18@pitt.edu>
remove unneeded


remove unneeded


Review comments
move changes entry up


clean up


clean up
